### PR TITLE
Allow service installations to be disabled

### PR DESCRIPTION
**Pull request by duncanhall**
This feature allows the future installation of services to be disabled, without affecting any current infrastructure.

selection_016

From the _Servers > ASG > Services_ view, individual services can be toggled between enabled or disabled. This sets a `Status` value in the key/value store for that particular combination of _environment, server role, service, version and slice_.

To complete this feature, updates need to be made to https://github.com/trainline/consul-deployment-agent that check the `Status` value of a service when any new infrastructure is created, and ignore installs for disabled services. Until this is done, the above work is behind a feature flag (`FEATURE_DISABLE_SERVICE`) that disables the feature by default.

To test this feature locally you can run:

```
FEATURE_DISABLE_SERVICE=true npm start
```

### CODE REVIEW

File: client/app/common/services/targetStateService.js [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+     var deferred = $q.defer();
+     var action = enable ? 'enable' : 'disable';
+     var url = '/api/services/'+ action + '/' + service.Name;
+     var data = {
+       slice: service.Slice,
+       environment: environment,
+       serverRole: role
+     };
+
+     $http.post(url, data)
```
Comment (filso) [inline client/app/common/services/targetStateService.js]: why not chain the promise? deferred doesn't seem necessary here

File: client/app/common/services/targetStateService.js [hunk 2]
```diff
@@ -0,0 +1,27 @@
+/* Copyright (c) Trainline Limited, 2016. All rights reserved. See LICENSE.txt in the project root for license information. */
+'use strict';
+
+angular.module('EnvironmentManager.common').factory('targetStateService', function($rootScope, $q, $http) {
+
+ return {
+   changeDeploymentStatus: function(enable, service, role, environment) {
```
Comment (filso) [inline client/app/common/services/targetStateService.js]: not necessary to change, but usually with setters / updaters / changers first you would specify target of action, and value (enable) would come last

File: client/app/environments/dialogs/DeployModalController.js [hunk 3]
```diff
@@ -21,8 +21,6 @@ angular.module('EnvironmentManager.environments').controller('DeployModalControl
     };
 
     function init() {
-      console.log('Initialising Deploy Controller for ' + parameters.Environment.EnvironmentName);
```
Comment (filso) [inline client/app/environments/dialogs/DeployModalController.js]: 👍

File: client/app/environments/dialogs/asg/asgInstances.js [hunk 4]
```diff
@@ -8,7 +8,7 @@ angular.module('EnvironmentManager.environments').component('asgInstances', {
     asgState: '<',
   },
   controllerAs: 'vm',
-  controller: function (awsService, $q, $uibModal) {
+  controller: function ($uibModal) {
```
Comment (filso) [inline client/app/environments/dialogs/asg/asgInstances.js]: 👍

File: server/modules/environment-state/getServiceTargetStates.js [hunk 5]
```diff
@@ -2,93 +2,78 @@
...
-      };
-    }
-  });
-}
-
-function getServiceOverallHealth(healthChecks, instances) {
-  if (_.some(healthChecks, { Status: Enums.HEALTH_STATUS.Error })) {
     return {
-      Status: Enums.HEALTH_STATUS.Error,
+      Name: checks[0].Name,
```
Comment (filso) [inline server/modules/environment-state/getServiceTargetStates.js]: 👍

File: server/modules/environment-state/getServiceTargetStates.js [hunk 6]
```diff
@@ -2,93 +2,78 @@
...
-    };
-  }
+  });
 }
 
-function* getServicesState(environmentName, runtimeServerRoleName, instances) {
-  let servicesList = yield serviceReporter.getServicesList(environmentName, runtimeServerRoleName);
+function getServiceOverallHealth(healthChecks) {
+  return {
+    Status:_.some(healthChecks, { Status: HEALTH_BAD }) ? HEALTH_BAD : HEALTH_GOOD
```
Comment (filso) [inline server/modules/environment-state/getServiceTargetStates.js]: 👍 , just formatting. We'll have also third state, WARNING, in which case logic will be more complex

File: server/modules/environment-state/getServiceTargetStates.js [hunk 7]
```diff
@@ -2,93 +2,78 @@
...
     });
 
     let serviceInstances = _.filter(instances, instance => _.some(instance.Services, { Name: service.Name }));
-
     let healthyNodes = _.filter(serviceInstances, (instance) => instance.OverallHealth.Status === Enums.HEALTH_STATUS.Healthy);
     let instancesHealthCount = healthyNodes.length + '/' + serviceInstances.length;
-
-    let serviceHealthChecks = getServiceChecksInfo(serviceObjects)
+    let serviceHealthChecks = getServiceChecksInfo(serviceObjects);
+    let Status = service.hasOwnProperty('Status') ? service.Status : 'Enabled';
```
Comment (filso) [inline server/modules/environment-state/getServiceTargetStates.js]: not sure if that's the best name for enabled / disabled service, status could also point at AWS instance status, might be confusing, especially if we add this to instance

Comment (duncanhall) [inline server/modules/environment-state/getServiceTargetStates.js]: Definitely open to changing this (particularly if the CDA team has other suggestions / requirements), but remember this is completely separate from AWS and instances, and doesn't ensure absence or presence of a service.

This flag is designed only to stop future installations of a service. Any services that are currently installed will be completely left alone, regardless of a change to the `Status` value (hence `Ensure: 'Absent'` would be misleading). 

In our Consul key/value store currently, services at this level look like:

```
{
  Name: "ServiceName",
  Version: "ServiceVersion",
  Slice: "blue|green|none",
  DeploymentId: "guid"
}
```
This new property gets added to the object above. Perhaps `InstallationStatus: Enabled | Disabled` might be better?

Comment (duncanhall) [inline server/modules/environment-state/getServiceTargetStates.js]: `Action: Ignore`

Comment (filso) [inline server/modules/environment-state/getServiceTargetStates.js]: `InstallationStatus` sounds good! indicates it's about (further) installation

**Pull request merged**